### PR TITLE
feat: Solana tokens

### PR DESCRIPTION
**Pull request by memoriesadrift**
## Description

This PR is the first in the Solana tokens feature planned for phase 2: it displays tokens in the wallet and adds them to account info.

We are *not* going to implement adding tokens ("importing tokens") in Solana, as the mechanism is very different from ETH and requires a transaction in order to create a new account on chain to hold the token. The majority of wallets in the Solana ecosystem do not allow the user to explicitly manage token accounts and the transfers of tokens to and from token accounts are handled automatically by the wallet. 

#### How token accounts work on Solana
The user's system program address (= main wallet address) is used to derived new accounts, one for each token. The new account's address can be used to receive that token and no other token.  
How most wallets handle this and how we plan to handle this in suite is that when a user wants to send tokens to another user, they simply use the recipient's system program address (= main wallet address) and while constructing the transaction the wallet checks if the recipient already has a token account for the given token or not.  
If the recipient has a token account, the wallet constructs a transaction sending the token to that address, whereas if there is not token address for that token an instruction to create one for the recipient is added to the transaction and is [funded by the sender](https://spl.solana.com/token#example-transferring-tokens-to-another-user-with-sender-funding).

## Related Issue

Resolve 

## Screenshots:

obrázok

### CODE REVIEW

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 1]
```diff
@@ -1,20 +1,29 @@
...
+
 const getAccountInfo = async (request: Request<MessageTypes.GetAccountInfo>) => {
     const { payload } = request;
     const api = await request.connect();
 
-    const accountInfo = await api.getAccountInfo(new PublicKey(payload.descriptor));
+    const publicKey = new PublicKey(payload.descriptor);
+
+    const accountInfo = await api.getAccountInfo(publicKey);
+    const tokenInfo = await api.getParsedTokenAccountsByOwner(publicKey, {
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/index.ts]: Please check [this commit](https://github.com/vacuumlabs/trezor-suite/pull/17/commits/e0845eaca09ff7222011c15868bb01b6be325f45) from the tx history PR, it gets transactions only if  `payload.details === txs`

We should get `tokens` only if payload.details === `tokens` or `tokenBalances`

Check how its done for other backends please

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
 import { MESSAGES, RESPONSES } from '@trezor/blockchain-link-types/lib/constants';
 import { Connection, PublicKey } from '@solana/web3.js';
+import { TokenInfo } from 'packages/blockchain-link-types/lib';
 
 export type SolanaAPI = Connection;
 
 type Context = ContextType<SolanaAPI>;
 type Request<T> = T & Context;
 
+const TOKEN_PROGRAM_PUBLIC_KEY = new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/index.ts:26]: Is this same for testnet and mainnet?

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 3]
```diff
@@ -14,6 +14,34 @@ type Request<T> = T & Context;
 
 const TOKEN_PROGRAM_PUBLIC_KEY = new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
 
+// This function matches the token mint to a name and symbol for some of the most popular tokens
+// Can be extended however desired, or replaced by a call to a token registry  API
+const getWhitelistedTokenInfo = (mint: string): { name: string; symbol: string } => {
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/index.ts]: I would put this into a separate file, e.g. `src/workers/solana/tokenList`

Also we talked about having and option to fetch the token list... if not do it now, I would leave a PR comment proposing what can be done => "we can fetch token metadata from here and parse them you guys would prefer"

**Review by PeterBenc — COMMENTED** [review]
Other than the comments, LGTM

I have some suggestions for the the PR description, but we can have a call about that, TDLR: I would try to explain the issue in a simpler terms

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 2 above)
Comment (memoriesadrift) [reply packages/blockchain-link/src/workers/solana/index.ts:26]: yes

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 1 above)
Comment (memoriesadrift) [reply packages/blockchain-link/src/workers/solana/index.ts]: still looking into this, as adding it caused the tokens to fail to load 😅

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts:1]: Ideally, this file would contain either the whitelist or logic for fetching the token metadata and it would be called in `getAccountInfo` passed to `transformToken` function which would then do the parsing. `transformToken` should be in `blockchain-utils` package, see the tx history for reference. At least this makes sense to me and it provides a good way to test everything.

Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts:1]: Also, I guess you had this plan anyway, but

Ideally split the commits into 
1. fetch tokens, do some basic parsing, without whitelist or tokenList
2. add whitelist
3. replace whitelist with tokenList

At least that's the way I would do it, since we want to bring attention to `nothing vs whitelist vs tokenList` and have the rest of the logic in separate commit.

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 1 above)
Comment (PeterBenc) [reply packages/blockchain-link/src/workers/solana/index.ts]: I guess we need to test it, but my logic is that tokens should be `undefined` when the `details` is `basic` and defined when its `tokens` or `tokenBalances`. But I am not sure whether this distinction is actually used or its just some leftover 

HAve you found out the difference between `tokens` and `token-balances` ?

Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/index.ts]: Pls move this line (fetching tokenInfo) after the `if(!accountInfo)` condition. IMO we do not need to fetch if the account is empty

File: packages/blockchain-link-utils/src/solana.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { A, F, pipe } from '@mobily/ts-belt';
+import { AccountInfo, ParsedAccountData, PublicKey } from '@solana/web3.js';
+import { TokenInfo } from 'packages/blockchain-link-types/lib';
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/solana.ts]: this import will fail, you need to do `@trezor/blockchain-link-types`

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        symbol: 'RAY',
+    },
+];
+const fetchSolanaTokenList = async (): Promise<
+    Array<{ mint: string; name: string; symbol: string }>
+> =>
+    Promise.resolve(TOKEN_WHITELIST);
+
+export const parseTokenInfo = async (
+    accountTokenQueryResult: { value: any } | undefined,
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts]: Better type pls

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        name: 'Raydium',
+        symbol: 'RAY',
+    },
+];
+const fetchSolanaTokenList = async (): Promise<
+    Array<{ mint: string; name: string; symbol: string }>
+> =>
+    Promise.resolve(TOKEN_WHITELIST);
+
+export const parseTokenInfo = async (
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts]: I would remove tihs fn and do the logic in `/solana/index`, the name of this function is `parse` even tho it also fetches.  I would keep the fetching in the `index` and pass stuff to `transformTokens`

File: packages/blockchain-link-utils/src/solana.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        ? { name: fetchedTokenInfo.name, symbol: fetchedTokenInfo.symbol }
+        : {
+              name: mint,
+              symbol: `${mint.slice(0, 3)}...`,
+          };
+};
+
+export const transformTokenInfo = (
+    tokenAccounts: TokenAccount[],
+    tokenNameAndSymbolMap: TokenDetailByMint,
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/solana.ts]: I would call this `tokenDetailByMint` as the type suggests

File: packages/blockchain-link-utils/src/solana.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { A, F, pipe } from '@mobily/ts-belt';
+import { AccountInfo, ParsedAccountData, PublicKey } from '@solana/web3.js';
+
+import { TokenInfo } from '@trezor/blockchain-link-types/lib';
+import { TokenDetailByMint } from '@trezor/blockchain-link-types/lib/solana';
+
+type TokenAccount = { account: AccountInfo<ParsedAccountData>; pubkey: PublicKey };
+
+export const getTokenNameAndSymbolFromWhitelist = (mint: string, tokenMap: TokenDetailByMint) => {
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/solana.ts]: As a reviewer, at this point I haven't seen no whitelist so the naming is a bit confusing, and also, it will become unrelavant in the last commit so I would stay away from the word `whitelist` as the implementation of this fn does not really care if the `tokenMap` came from whitelist or not

Also `tokenDatailsByMint` would be a better name for that argument

File: packages/blockchain-link-utils/src/solana.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    const fetchedTokenInfo = tokenMap[mint];
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/solana.ts]: Here we call this `fetchedTokenInfo` even tho fn name says whitelist. Also at this stage nothing is fetched, nor this fn cares about that.

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 10]
```diff
@@ -35,6 +41,18 @@ const getAccountInfo = async (request: Request<MessageTypes.GetAccountInfo>) =>
...
     }
 
+    const tokenInfo = await api.getParsedTokenAccountsByOwner(publicKey, {
+        programId: TOKEN_PROGRAM_PUBLIC_KEY,
+    });
+
+    // Fetch token info only if the account owns tokens
+    let tokens: TokenInfo[] = [];
+    if (tokenInfo.value.length > 0) {
+        const tokenNameAndSymbolList = await fetchSolanaTokenList();
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/index.ts]: This naming is IMO also quite strange, maybe
`const tokensMetadata = await getSolanaTokensMetadata()`

File: packages/blockchain-link-utils/src/tests/fixtures/solana.ts [hunk 11]
```diff
@@ -1,5 +1,64 @@
+import { PublicKey } from '@solana/web3.js';
+
+const tokenWhitelist = {
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/tests/fixtures/solana.ts]: put these tests in the end so its obvious that these tests are unrelated to `whitelist vs fetchedTokenList`, also `tokensMetadata` naming would allow us to avoid using `whitelist`

File: packages/blockchain-link-utils/src/tests/fixtures/solana.ts [hunk 12]
```diff
@@ -12,4 +71,23 @@ export const fixtures = {
             },
         },
     ],
+    transformTokenInfo: [
+        {
+            description: 'parses token info from token accounts api response',
+            input: {
+                accountInfo: tokenAccountInfo,
+                map: tokenWhitelist,
```
Comment (PeterBenc) [inline packages/blockchain-link-utils/src/tests/fixtures/solana.ts]: `map: tokenWhitelist` is a strange naming IMO

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 13]
```diff
@@ -1,3 +1,19 @@
...
+        name: 'Tether',
+        symbol: 'USDT',
+    },
+    '4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R': {
+        name: 'Raydium',
+        symbol: 'RAY',
+    },
+};
+
+export const fetchSolanaTokenList = async (): Promise<TokenDetailByMint> =>
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts]: Here again, its strange that the function is called `fetch` even tho it does not fetch anything, we might end up dropping the token list and this will stay as it is in this commit, and this `fetch` will never be relevant. I would call this `getTokensMetadata` and let the fetching or getting the whitelist be an implementation detail of this fn

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 14]
```diff
@@ -1,19 +1,25 @@
...
+    'https://cdn.jsdelivr.net/gh/viaprotocol/tokenlists/all_tokens/solana.json';
+
+export const fetchSolanaTokenList = async (): Promise<TokenDetailByMint> => {
+    const tokenListResult: { address: string; name: string; symbol: string }[] = await (
+        await fetch(solanaTokenListUrl)
+    ).json();
 
-export const fetchSolanaTokenList = async (): Promise<TokenDetailByMint> =>
-    Promise.resolve(TOKEN_WHITELIST);
+    const tokenMap = tokenListResult.reduce(
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts:15]: NP: just naming, please if we create an object of type `TokenDetailByMint` if possible lets call it `TokenDetailByMint`

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+import { TokenDetailByMint } from '@trezor/blockchain-link-types/lib/solana';
+
+export const fetchSolanaTokenList = async (): Promise<TokenDetailByMint> => Promise.resolve({});
```
Comment (PeterBenc) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts]: Since we introduced this in the first commit, I would leave a PR comment here saying 
`The two alternatives of implementing this function are shown in the next two commits` or something like that

**Review by PeterBenc — COMMENTED** [review]
Basically LGTM, is just readability remarks, mostly consistency of naming

Feel free to come up with different naming if what I suggested is not ideal, just keep it consistent and avoid situations where the naming is suggesting something that's not true
